feat: tolgee platform client

stepan662 opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  - Introduced the @tginternal/client package: a typed API client with API-key and user-token auth, improved error messages, utilities for key info and unresolved-translation reporting, and a consolidated public API.
  - Automated publish workflow to release the client package on demand and after releases.

* **Chores**
  - Added build/config tooling, version generation, and package manifest for the client.
  - Expanded CI to include client build/tests.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

---

## Additions on top of the original client package

This branch also lands **typed webhooks** so anything consuming `@tginternal/client` (apps, plugins, custom webhook receivers) can type-check webhook payloads end-to-end without hand-maintaining shapes.

### Backend — OpenAPI 3.1 `webhooks` section

- `application.yaml` now emits **OpenAPI 3.1** (`springdoc.api-docs.version=openapi_3_1`).
- New `AppWebhookOpenApiCustomizer` injects a `webhooks` entry per visible `ActivityType` into the "All Internal" group. Per-event payload schemas honour the three size-control flags on `ActivityType`:
  - `onlyCountsInList=true` → `modifiedEntities` is constrained to `{}`; signal moves to `counts` and `params`.
  - `restrictEntitiesInList=[X::class, …]` → `modifiedEntities` is restricted to those entity classes.
  - `paramsProvider=…` → a typed `params` property appears on `activityData`.
- Per-entity `<EntityClass>Modifications` schemas (`TranslationModifications`, `KeyModifications`, …) are reflectively built from `@ActivityLoggedProp` annotations on each `@ActivityLoggedEntity`.
- `OpenApiUnusedSchemaCleaner` now also walks `openAPI.webhooks` so synthesized payload schemas survive the cleanup pass.

### Tests

- **`WebhookSpecGenerationTest`** — 214 parameterized cases prove the *spec* is well-formed: every visible `ActivityType` has a webhook entry with a `$ref` to a payload schema, payload schemas honour the three flag-driven shape rules, and per-entity Modifications schemas mirror their `@ActivityLoggedProp` annotations.
- **`WebhookContractFixture`** + 4 contract test classes — prove the *wire payload* matches the spec. Pattern: per-event test class with one trigger method calling the real controller/service path, then `fixture.waitForWebhookWithType(event)` + `fixture.assertConformsTo(payload, event)`. **No abstract base class** (team convention) — every helper is called via an explicit `fixture.X(…)` so origin is grep-able from the call site.
  - `SetTranslationsWebhookContractTest` — full-data, single entity.
  - `KeyNameEditWebhookContractTest` — full-data + asserts the `modifications.name.{old,new}` diff.
  - `CreateLanguageWebhookContractTest` — full-data, Language only.
  - `DeleteLanguageWebhookContractTest` — proves `restrictEntitiesInList = [Language::class]` is honoured even when the cascade deletes translations.

### Client (`@tginternal/client`)

- `openapi-typescript` is already on `^7.9.1` here (vs the webapp's 5.x), so the generated `src/schema.generated.ts` carries the `webhooks` namespace automatically.
- README documents the consumer pattern:
  ```ts
  import type { webhooks } from "@tginternal/client";
  type SetTranslationsPayload =
    webhooks["SET_TRANSLATIONS"]["post"]["requestBody"]["content"]["application/json"];
  ```
- Discriminated-union usage via `activityData.type` is documented too, plus the HMAC verification recipe for the `Tolgee-Signature` header.

### Deferred (clean follow-ups)

- Contract tests for `IMPORT` (counts-only — needs `ImportService.import(...)` driver), `BATCH_MACHINE_TRANSLATE` (counts-only + `paramsProvider` — `@Import(BatchJobBaseConfiguration::class)` interacts oddly with `@MockitoBean` wiring; needs its own debugging session), and `BRANCH_MERGE` (EE-only). The pattern is established — each is a copy-paste-and-swap of an existing class.
- Webapp's `openapi-typescript` 5.4.2 → 7.x bump (~40 `useApiQuery`/`useApiMutation` call sites need updates due to operation-type shape changes). Webapp's 5.x still tolerates 3.1 input for the schema it consumes; bump can ship as its own scope.

### Versions published

- `@tginternal/client@0.0.0-prerelease.c8329b552` (workflow_dispatch on this branch)
- Additional docs commit on top — next `workflow_dispatch` will publish `…4ca8a9a4`.

dkrizan commented on client/src/ApiClient.ts:
```diff
@@ -0,0 +1,174 @@
+export function createApiClient(props: ApiClientProps) {
+  let {
+    projectId,
+    baseUrl = "https://app.tolgee.io",
```
this could be in `constants.ts` too

dkrizan reviewed (commented):
Code looks good! Left a few minor notes.

zomp commented on client/tsconfig.json:
```diff
@@ -0,0 +1,30 @@
+    "noEmit": false,
+    "outDir": "./lib"
+  },
+  "include": ["src"]
```
Does this work? Should not it be `"src/**/*"`?

JanCizmar committed:
feat: tolgee platform client

Adds a new internal TypeScript client package (`@tginternal/client`)
that wraps openapi-fetch with typed paths from the platform's OpenAPI
schema, plus helpers for API-key parsing, login, and friendly error
messages. The package is versioned in lockstep with the platform and
published via a dedicated workflow (latest tag on release, prerelease
tag tied to the git short SHA otherwise).

Highlights:
- ApiClient with PAK/PAT/user-token auth, optional auto-throw, verbose
  HTTP logging, login helper, and a getApiKeyInformation that hits
  `/v2/api-keys/current` or `/v2/pats/current` depending on the key
  prefix.
- errorFromLoadable maps backend status codes and unresolved-conflict
  payloads to readable error messages.
- Schema is regenerated from the running backend during publish; the
  generated file lives at `client/src/schema.generated.ts` and is
  committed to keep type-checks reproducible without a live backend.
- New `client-checks` job in `test.yml` runs build + tests on every
  PR.

Co-Authored-By: Štěpán Granát <granat.stepan@gmail.com>

JanCizmar committed:
chore: switch client publish to npm trusted publishing (OIDC)

Adopts the same pattern as tolgee-js: authenticate to npm via GitHub
Actions OIDC instead of a long-lived NPM_TOKEN secret.

- Add top-level `permissions: id-token: write, contents: read` so the
  job can mint an OIDC token.
- Configure `actions/setup-node` with `registry-url` so the npm CLI
  knows where to redeem the OIDC token.
- Drop the `~/.npmrc` echo step and the `NPM_TOKEN` env vars on both
  publish steps.
- Add `--provenance` to both `npm publish` invocations so the published
  tarballs carry build provenance attestations.

Note: requires the `@tginternal/client` package on npmjs.org to be
configured with a Trusted Publisher pointing at
`tolgee/tolgee-platform` + `publish-client.yml`.

JanCizmar committed:
chore: bump publish workflow Node to 24 for npm trusted publishing

npm trusted publishing requires npm 11.5.1+ (July 2025). Node 22.x
ships with npm 10.x and only attempts OIDC for provenance, not for
the publish PUT itself — which kept failing with a 404 ("permission
denied"). Node 24.x ships with npm 11.x and supports OIDC end-to-end.

Matches the Node version used in tolgee-js's release workflow.

JanCizmar committed:
chore: add repository, homepage, bugs, license to client package.json

npm provenance attestation requires `repository.url` to match the
GitHub repo the OIDC token was issued from. Without it, publish fails
with 422 ("Failed to validate repository information"). Added
`repository`, plus `homepage`, `bugs`, and `license` while here —
these are required to publish provenance-attested tarballs and
populate the npm package page.

JanCizmar committed:
feat: typed webhooks via OpenAPI 3.1 webhooks section

- Springdoc emits OpenAPI 3.1 (springdoc.api-docs.version=openapi_3_1).
- AppWebhookOpenApiCustomizer injects a `webhooks` entry per visible
  ActivityType into the "All Internal" group's spec. Per-event
  AppWebhookPayload_<TYPE> schemas honour the three size-control flags on
  ActivityType: onlyCountsInList collapses modifiedEntities to {};
  restrictEntitiesInList narrows the allowed entity-class keys;
  paramsProvider surfaces a typed params property.
- Per-entity Modifications schemas (e.g. TranslationModifications,
  KeyModifications, LanguageModifications) are reflectively built from
  @ActivityLoggedProp annotations; non-annotated entities fall back to a
  permissive additionalProperties: PropertyModification shape so $refs
  resolve.
- Extended OpenApiUnusedSchemaCleaner to also walk openAPI.webhooks when
  collecting referenced schemas, otherwise the cleaner would strip our
  synthesized payload schemas.
- WebhookSpecGenerationTest covers 214 parameterized cases: 3.1 advertised,
  webhooks entry exists for every visible ActivityType, schema $refs land
  on existing components, counts-only/restricted/paramsProvider variants
  shape correctly, per-entity Modifications mirror the annotations.

Deferred to follow-up commits / scopes:
- WebhookPayloadSchemaContractTest (drives real activities, captures
  payloads, validates against schemas). Foundation is here; per-event
  trigger methods + parameterized matrix are next.
- Bumping webapp/openapi-typescript from 5.4.2 to 7.x. The 7.x typing of
  paths/operations breaks ~40 webapp useApiQuery/useApiMutation call sites
  because the hook helpers infer differently; non-trivial migration that
  shouldn't gate the typed-webhooks shipment. client/ already on 7.9.1.
- client/src/schema.generated.ts regeneration + vitest smoke test —
  requires running backend; plugin authors get the types as soon as
  someone runs `cd client && TOLGEE_API_PORT=8080 npm run schema` against
  a backend with these changes.

JanCizmar committed:
test: webhook payload contract tests across representative shape categories

- WebhookContractFixture: injected @Component (no abstract base), owns the
  WebhooksTestData fixture, mocks webhookRestTemplate, captures the JSON
  Tolgee POSTs, waits for the right activity type via
  waitForWebhookWithType(eventName), and asserts the captured payload
  conforms to the OpenAPI spec's webhook contract via assertConformsTo().
  Conformance is checked structurally (top-level fields, activityData.type
  discriminator, modifiedEntities keys ⊆ spec-declared keys) — the
  WebhookSpecGenerationTest already proves the spec shape.
- SetTranslationsWebhookContractTest: pre-creates the key (so the cascade
  doesn't pollute modifiedEntities) then sets the translation; asserts the
  payload has only Translation entries.
- KeyNameEditWebhookContractTest: creates a key, renames it; asserts the
  modifications.name.{old,new} diff is present.
- CreateLanguageWebhookContractTest: creates a language; asserts only
  Language entries.
- DeleteLanguageWebhookContractTest: creates and deletes a language;
  asserts modifiedEntities is constrained to Language (proves
  restrictEntitiesInList = [Language::class] is honoured on the wire).

Pattern is grep-able and extension-friendly: copy a test class, change the
event name + trigger method. Coverage for IMPORT (onlyCountsInList),
BATCH_MACHINE_TRANSLATE (onlyCountsInList + paramsProvider), and
BRANCH_MERGE (EE) is deferred — those need real ImportService /
BatchJobService / BranchMergeService driver setup that's worth its own
focused commit.

JanCizmar committed:
docs(client): document the webhooks namespace for receiver typing

Plugin authors building webhook receivers can import per-event payload
types from the generated schema (e.g. webhooks["SET_TRANSLATIONS"]) and
discriminate via activityData.type. Document both the single-event and
discriminated-union patterns, plus the HMAC signature verification recipe.

dkrizan committed:
chore: regenerate webapp API schema for typed webhooks

dkrizan committed:
Merge remote-tracking branch 'origin/main' into stepangranat/tolgee-client-lib

dkrizan committed:
chore: regenerate webapp API schema after merging main

dkrizan committed:
style: apply ktlint formatting to webhook OpenAPI code
